Cabal/Distribution/InstalledPackageInfo.hs

```diff
@@ -249,6 +252,26 @@ parseExposedModules :: Parse.ReadP r [ExposedModule]
 parseExposedModules = parseOptCommaList parse
 
 -- -----------------------------------------------------------------------------
+-- ABI dependency
+
+data AbiDependency = AbiDependency {
+        depUnitId  :: UnitId,
+        depAbiHash :: AbiHash
+    }
+  deriving (Eq, Generic, Read, Show)
+
+instance Text AbiDependency where
+    disp (AbiDependency uid abi) =
+        disp uid <<>> Disp.char '@' <<>> disp abi
+    parse = do
+        uid <- parse
+        _ <- Parse.char '@'
```

Comment:
Minor nit-pick. I think '@' has the wrong connotation. It's as if you're selecting some sub-revision of something, but that's not the correct interpretation. It's really like an additional checksum; it caches you when you do something obviously bogus (but isn't fool proof). So I'd suggest a different joiner syntax like '&' or '+' or ' abi '.